File: app/src/main/java/tech/diarmaid/koohiiaite/adapter/KanjiListFilterAdapter.java
package tech.diarmaid.koohiiaite.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import org.honorato.multistatetogglebutton.MultiStateToggleButton;
import org.honorato.multistatetogglebutton.ToggleButton;

import tech.diarmaid.koohiiaite.R;
import tech.diarmaid.koohiiaite.activity.KanjiListActivity;

/**
 * Adapter for the spinner in the kanji list
 */
public class KanjiListFilterAdapter extends ArrayAdapter<String> {

    private Context mContext;
    private LayoutInflater layoutInflater;

    private FilterState joyoFilter = FilterState.UNSET;
    private FilterState keywordFilter = FilterState.UNSET;
    private FilterState storyFilter = FilterState.UNSET;

    public KanjiListFilterAdapter(Context context, int resource, String[] objects) {
        super(context, resource, objects);
        layoutInflater = LayoutInflater.from(context);
        mContext = context;
    }

    @Override
    public View getDropDownView(int position, View cnvtView, @NonNull ViewGroup prnt) {
        return getCustomView(position, cnvtView, prnt);
    }

    @Override
    public View getView(int pos, View cnvtView, @NonNull ViewGroup prnt) {
        TextView text = new TextView(getContext());
        text.setText(R.string.filter_hint);
        return text;
    }

    public View getCustomView(int position, View convertView,
                              ViewGroup parent) {
        ViewHolderItem viewHolder;
        if (convertView == null) {
            convertView = layoutInflater.inflate(R.layout.spinner_filter, parent, false);
            viewHolder = new ViewHolderItem();
            viewHolder.joyoButton = convertView.findViewById(R.id.toggle_joyo);
            viewHolder.keywordButton = convertView.findViewById(R.id.toggle_keyword);
            viewHolder.storyButton = convertView.findViewById(R.id.toggle_story);
            viewHolder.joyoButton.setValue(joyoFilter.getStateNum());
            viewHolder.keywordButton.setValue(keywordFilter.getStateNum());
            viewHolder.storyButton.setValue(storyFilter.getStateNum());

            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolderItem) convertView.getTag();
        }

        viewHolder.joyoButton.setOnValueChangedListener(
                new ToggleButton.OnValueChangedListener() {
                    @Override
                    public void onValueChanged(int i) {
                        joyoFilter = FilterState.getStateFor(i);
                        if (mContext instanceof KanjiListActivity) {
                            ((KanjiListActivity) mContext).notifyFilterChanged();
                        }
                    }
                });

        viewHolder.keywordButton.setOnValueChangedListener(
                new ToggleButton.OnValueChangedListener() {
                    @Override
                    public void onValueChanged(int i) {
                        keywordFilter = FilterState.getStateFor(i);
                        if (mContext instanceof KanjiListActivity) {
                            ((KanjiListActivity) mContext).notifyFilterChanged();
                        }
                    }
                }
        );

        viewHolder.storyButton.setOnValueChangedListener(
                new ToggleButton.OnValueChangedListener() {
                    @Override
                    public void onValueChanged(int i) {
                        storyFilter = FilterState.getStateFor(i);
                        if (mContext instanceof KanjiListActivity) {
                            ((KanjiListActivity) mContext).notifyFilterChanged();
                        }
                    }
                }
        );

        return convertView;
    }

    public FilterState getJoyoFilter() {
        return joyoFilter;
    }

    public void setJoyoFilter(Integer value) {
        joyoFilter = FilterState.getStateFor(value);
    }

    public FilterState getKeywordFilter() {
        return keywordFilter;
    }

    public void setKeywordFilter(Integer value) {
        keywordFilter = FilterState.getStateFor(value);
    }

    public FilterState getStoryFilter() {
        return storyFilter;
    }

    public void setStoryFilter(Integer value) {
        storyFilter = FilterState.getStateFor(value);
    }

    public enum FilterState {
        UNSET(0),
        YES(1),
        NO(2);

        private final int stateNum;

        FilterState(int stateNum) {
            this.stateNum = stateNum;
        }

        public static FilterState getStateFor(int stateNum) {
            for (FilterState state : values()) {
                if (state.stateNum == stateNum) {
                    return state;
                }
            }
            Log.e("FilterState", "Couldn't find given filter state : " + stateNum);
            return UNSET;
        }

        public int getStateNum() {
            return stateNum;
        }
    }

    static class ViewHolderItem {
        MultiStateToggleButton joyoButton;
        MultiStateToggleButton keywordButton;
        MultiStateToggleButton storyButton;
    }
}

File: app/src/main/java/tech/diarmaid/koohiiaite/adapter/KanjiListFilterAdapter.kt
package tech.diarmaid.koohiiaite.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.google.android.material.button.MaterialButtonToggleGroup
import tech.diarmaid.koohiiaite.R
import tech.diarmaid.koohiiaite.activity.KanjiListActivity
import tech.diarmaid.koohiiaite.enumeration.FilterState

/**
 * Adapter for the spinner in the kanji list
 */
class KanjiListFilterAdapter(private val mContext: Context, resource: Int, objects: Array<String>) : ArrayAdapter<String>(mContext, resource, objects) {
    private val layoutInflater: LayoutInflater = LayoutInflater.from(mContext)

    var joyoFilter = FilterState.UNSET
    var keywordFilter = FilterState.UNSET
    var storyFilter = FilterState.UNSET

    override fun getDropDownView(position: Int, cnvtView: View?, prnt: ViewGroup): View {
        return getCustomView(cnvtView, prnt)
    }

    override fun getView(pos: Int, cnvtView: View?, prnt: ViewGroup): View {
        val text = TextView(context)
        text.setText(R.string.filter_hint)
        return text
    }

    private fun getCustomView(aConvertView: View?, parent: ViewGroup): View {
        var convertView = aConvertView
        val viewHolder: ViewHolderItem
        if (convertView == null) {
            convertView = layoutInflater.inflate(R.layout.spinner_filter, parent, false)
            viewHolder = ViewHolderItem()
            viewHolder.joyoGroup = convertView!!.findViewById(R.id.toggle_joyo)
            viewHolder.keywordGroup = convertView.findViewById(R.id.toggle_keyword)
            viewHolder.storyGroup = convertView.findViewById(R.id.toggle_story)

            // Set initial states
            setToggleGroupState(viewHolder.joyoGroup, joyoFilter.stateNum)
            setToggleGroupState(viewHolder.keywordGroup, keywordFilter.stateNum)
            setToggleGroupState(viewHolder.storyGroup, storyFilter.stateNum)

            convertView.tag = viewHolder
        } else {
            viewHolder = convertView.tag as ViewHolderItem
        }

        viewHolder.joyoGroup!!.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked) {
                val state = when (checkedId) {
                    R.id.toggle_joyo_all -> FilterState.UNSET
                    R.id.toggle_joyo_yes -> FilterState.YES
                    R.id.toggle_joyo_no -> FilterState.NO
                    else -> FilterState.UNSET
                }
                joyoFilter = state
                if (mContext is KanjiListActivity) {
                    mContext.notifyFilterChanged()
                }
            }
        }

        viewHolder.keywordGroup!!.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked) {
                val state = when (checkedId) {
                    R.id.toggle_keyword_all -> FilterState.UNSET
                    R.id.toggle_keyword_yes -> FilterState.YES
                    R.id.toggle_keyword_no -> FilterState.NO
                    else -> FilterState.UNSET
                }
                keywordFilter = state
                if (mContext is KanjiListActivity) {
                    mContext.notifyFilterChanged()
                }
            }
        }

        viewHolder.storyGroup!!.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked) {
                val state = when (checkedId) {
                    R.id.toggle_story_all -> FilterState.UNSET
                    R.id.toggle_story_yes -> FilterState.YES
                    R.id.toggle_story_no -> FilterState.NO
                    else -> FilterState.UNSET
                }
                storyFilter = state
                if (mContext is KanjiListActivity) {
                    mContext.notifyFilterChanged()
                }
            }
        }

        return convertView
    }

    fun setJoyoFilter(value: Int?) {
        joyoFilter = FilterState.getStateFor(value!!)
    }

    fun setKeywordFilter(value: Int?) {
        keywordFilter = FilterState.getStateFor(value!!)
    }

    fun setStoryFilter(value: Int?) {
        storyFilter = FilterState.getStateFor(value!!)
    }

    private fun setToggleGroupState(group: MaterialButtonToggleGroup?, state: Int) {
        if (group == null) return
        val buttonId = when (state) {
            0 -> group.getChildAt(0).id // All
            1 -> group.getChildAt(1).id // Yes
            2 -> group.getChildAt(2).id // No
            else -> group.getChildAt(0).id
        }
        group.check(buttonId)
    }

    internal class ViewHolderItem {
        var joyoGroup: MaterialButtonToggleGroup? = null
        var keywordGroup: MaterialButtonToggleGroup? = null
        var storyGroup: MaterialButtonToggleGroup? = null
    }
}
